### PR TITLE
Update menubutton examples: High Contrast Support and APG Coding Practices

### PR DESCRIPTION
**Pull request by jongund**
Updated menu button examples to:
1. Improve Javascript coding and use `key` rather than `keyCode` in keyboard events, and a single object
1. Improved high contrast support 

### Preview links

[Review: Menu Button for Links](https://raw.githack.com/w3c/aria-practices/issue1399-menubutton-updates/examples/menu-button/menu-button-links.html)
[Review: Menu Button for Actions](https://raw.githack.com/w3c/aria-practices/issue1399-menubutton-updates/examples/menu-button/menu-button-actions.html)
[Review: Menu Button for Actions using `aria-activedescendant`](https://raw.githack.com/w3c/aria-practices/issue1399-menubutton-updates/examples/menu-button/menu-button-actions-active-descendant.html)

### Review checklist

* [x] [Editorial review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#editorial_review) by @mcking65
* [x] [Functional review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#functional_review) by @curtbellew
* [x] [Visual design review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#design_review) by @a11ydoer
* [x] [Accessibility review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#accessibility_review) by @charmarkk
* [x] [Code review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#code_review) by @spectranaut (NOTE: there is a new requirement in the [code review process](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#code-review))
* [x] [Test review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#test_review) by @spectranaut

### CODE REVIEW

**Comment by jongund** [comment]
@zcorpan 

Can you look at the test cases that are failing for the updated menu button examples?
I can't figure out why they are failing, I get passing results for most of the tests and inconsistent failures on my local MacOS and Windows 10 machine.

1. macOS: links and aria-activedescent example all tests pass, page down, page up and end fail for actions with focus example.
1. Windows 10: both action examples pass and there about six failures related to keystrokes for the links example.

This is driving me nuts, what don't I understand about the testing framework?

**Comment by mcking65** [comment]
@jongund, I meant to ask you this on an earlier PR and then forgot ... in this description of one of the HCM support features:

>Focus is highlighted by adding and removing a border around the menu item with focus. 

Why is "and removing" included in the description? Isn't the border added when the element receives focus and then removed after it loses focus? It is not removed while the element has focus is it?

This text doesn't really describe how this border supports high contrast. I imagine that plenty of people add borders to elements with focus that don't help when users turn on high contrast in the operating system. So, what about this border makes it effective when operating system HC settings are enabled?

**Comment by jongund** [comment]
@mcking65 
The adding and removing border means adjusting both the width of the border and the padding for graphical rendering stability.  Earlier APG examples just made the border transparent when an item doesn't have focus , but that does not work in high contrast mode, which makes even transparent borders visible.  So you need to set the border to 0 width and then add the width to the padding to make the graphic rendering stable for items that do not have focus.

**Comment by mcking65** [comment]
@spectranaut 
@jongund 
WRT the tests, is there something worth capturing in our test writing documentation here?

**Comment by jongund** [comment]
@mcking65 
@spectranaut 
I think there is some worthwhile lessons on writing tests.  One thing I have noticed is the use of "click" events to open menus and this seems to have effects of moving the mouse over some part of the widget, which sometimes changes the behavior of the widget or the test (not sure).  So one suggestion might be to only using "sendKeys" when testing keyboard features.  I have changed a "click" to "sendKeys" in the menu button for actions example and regression test started working on Travis CI when it didn't before, even though the test passed on my local machine using the "click" approach.  It has certainly been an adventure learning about AVA for me, but I think I am starting to get the hang of it.

**Comment by jongund** [comment]
@charmarkk 
Mark thank you for the review, I am not sure there is a way to have a border in high contrast mode and use the border for indicating keyboard focus.  If this is important we will need to move to using the `outline` property, but personally I don't think we need the border in HCM.

**Comment by mcking65** [comment]
@jongund would you please delete the comments in the test files that Carolyn suggested are no longer needed?

@charmarkk Per your earlier comment regarding button border, would you please re-review with approval?

**Comment by mcking65** [comment]
@jongund,

Now that we have clarified the first point in accessibility features documentation about HCM, I'd like to better understand the second one:

>The arrow icons indicating the current state of expandable menus use in-line SVG images with the CSS currentColor  value for the fill  and stroke properties of the polygon  element. 

That states what is in the CSS, but not why? What is important here? Is it use of current-color in fill and stroke? What would fail if you didn't do this? How would it fail? Why would it fail?

**Comment by jongund** [comment]
@mcking65 
@carmacleod 
The use of `currentColor` changes the color of the arrow to use the same color as the text color value, so when a high contrast mode changes the text color, the SVG stroke and fill use the same color as the text.  Otherwise, when high contrast mode is enabled we get black or light gray arrows on a black background and there is little or no color contrast to see the arrows.

**Review by spectranaut — APPROVED** [review]
Thanks for addressing those code comments! Code review approval.

**Comment by mcking65** [comment]
To explain the benefit of and further clarify the second HCM feature, I have changed the list item that said:

>The arrow icons indicating the current state of expandable menus use in-line SVG images with the CSS currentColor  value for the fill  and stroke  properties of the polygon  element. 

To:

>To ensure the arrow icons that indicate the expanded or collapsed state of menus have sufficient contrast when high contrast settings invert colors to light text on a dark background, the color of the arrow icons changes to match the text color via use of the CSS currentColor  value for the fill  and stroke  properties of the SVG polygon  element.

**Comment by carmacleod** [comment]
@mcking65 
Your clarification of the HCM feature makes sense, except for the phrase "invert colors to light text on a dark background".
I think that phrase should be replaced with "are used", so that the sentence reads:

> To ensure the arrow icons that indicate the expanded or collapsed state of menus have sufficient contrast when high contrast settings are used, the color of the arrow icons changes to match the text color via use of the CSS currentColor value for the fill and stroke properties of the SVG polygon element.

HCM can be light on dark, or dark on light, or a set of customizable colors in between. We should test both "High Contrast Dark" and "High Contrast Light" (in Edge and FF) to ensure that we've caught all of the places where either a dark color or a light color was hardcoded.

**Comment by carmacleod** [comment]
@jongund I'm not sure about having a menu button with no visible border in high contrast mode? The only visual affordance is the adjacent drop-down triangle. It might be a bit of a cognitive leap for someone to realize it's a button?

I tested some plain (i.e. no extra css) [html buttons](https://carmacleod.github.io/playground/button-tests.html) in HCM to see what they do. They have a thin outline, which turns into a slightly thicker outline when they are focused. This looks ok in HCM Light and HCM Dark **in Edge**. Firefox, however, uses the "text selection" background color for focused buttons but they don't invert the foreground button text color, so focused buttons end up with very poor contrast. Pretty sure this is a Firefox HCM bug.

I won't hold up this PR over the button not having a border, because I don't know the right answer. It seems that people do have buttons without borders in some designs (particularly in toolbars), but the borderless buttons seem to typically be icon-only buttons, not text-only or text-plus-icon buttons. For example, on this github page, the editing toolbar buttons (Bold, Italic, Add link, etc) do not have a border, but the "Merge pull request", "Close with comment", and "Comment" buttons do have a border. I think we should go ahead and merge this PR, then open a new issue and ask for help from folks who have used and designed for high contrast.

PS: Firefox also has a 1-pixel vertical jiggle when hovering or focusing the menu button. Probably a Firefox bug as well - I'm only mentioning it in case you happen to know of a workaround that doesn't cause jiggle in the other browsers. This FF jiggle is not worth holding up the merge over, either.

**Review by carmacleod — APPROVED** [review]
+1

**Comment by jongund** [comment]
@carmacleod 
Thank you for the feedback and suggestions.  I will add a 1 pixel border and then changing it to a 2 or 3 pixel border when the button has focus.  I will also look into the giggle issue.

**Comment by jongund** [comment]
@carmacleod 
@charmarkk 
I have added a 1 pixel border on the button and it changes to 3 pixels with focus.  So can you please review.
Also Carolyn I am not seeing the gigglw in Firefox, can you help me understand what actions cause the giggle and what is giggling?
Thanks,
Jon

**Review by carmacleod — APPROVED** [review]
Wow, @jongund - thank-you! Now all the menu buttons are very visible, in both High Contrast Black and High Contrast White!

Regarding the vertical jiggle, here's the easiest way to see it:
1. In Firefox, open any of the menu button examples (it's easiest to see with one of the [Action menu button](https://raw.githack.com/w3c/aria-practices/issue1399-menubutton-updates/examples/menu-button/menu-button-actions.html) examples)
2. Scroll the page so that the Action menu button is in the middle of the page
3. Make sure the menu button does not have focus
4. Move the mouse over the button, and then move it away (repeat. mouse in, mouse out, mouse in, mouse out)
5. Observe that all of the page content below the menu button (i.e. the Last Action field and the Accessibility Features section) moves up one pixel on mouse hover, and then back down one pixel on mouse out.

Good luck. :)
This may be a Firefox bug. Not sure.

**Comment by mcking65** [comment]
@carmacleod wrote:

> Your clarification of the HCM feature makes sense, except for the phrase "invert colors to light text on a dark background".
> I think that phrase should be replaced with "are used", so that the sentence reads:
> 
> > To ensure the arrow icons that indicate the expanded or collapsed state of menus have sufficient contrast when high contrast settings are used, the color of the arrow icons changes to match the text color via use of the CSS currentColor value for the fill and stroke properties of the SVG polygon element.

Somehow, with this wording, it feels non-obvious to me that the consequence of not using currentColor is that the dark arrows disappear into the dark background. To make the guidance easy to understand, I think this is the point we need to emphasize.

Maybe this would be better:

> To ensure the darkly colored arrow icons that indicate the expanded or collapsed state of menus do not disappear into a dark background when high contrast settings are used to make the background dark and the forground light, the CSS currentColor value is used to set the icon color so that its color will change to match the text color.

Does this work?

**Comment by carmacleod** [comment]
How about:
> To ensure the arrow icons that indicate the expanded or collapsed state of menus do not disappear into the background when high contrast settings are used, the color of the arrow icons changes to match the text color via use of the CSS currentColor value for the fill and stroke properties of the SVG polygon element.

My point was that high contrast settings are not only used to make the background dark and the foreground light... they are also used to make the background light and the foreground dark. Our icons are actually both light and dark, even without high contrast mode. Focus/hover switch them from one to the other.

I do think it's important to keep the part about currentColor being used for the fill and stroke of the SVG.

**Comment by jongund** [comment]
@carmacleod 
@mcking65 
How about:
To ensure the arrow icons used to indicate the expanded or collapsed state have sufficient contrast with the background when high contrast settings invert colors, the CSS `currentColor` value for the `fill` and `stroke` properties of the SVG `polygon` element are used to synchronize the color with text content.  If specific colors are used to specify the `fill` and `stroke` properties, in high contrast mode these colors will remain the same and in many cases will not have sufficient color contrast with the background, and in some cases will be invisible when they have the same color as the background.

**Comment by jongund** [comment]
@carmacleod 
I was not able to replicate the Firefox jiggle issue on either my Mac or Windows machine, so not sure what t do at this point with this issue.

**Comment by carmacleod** [comment]
@jongund Ignore the jiggle. :)
I like your latest words for the currentColor explanation. That makes it quite clear.

**Comment by jongund** [comment]
@carmacleod 
I updated the documentation and it is available in the pull request.

**Review by curtbellew — APPROVED** [review]
Ran through some functional testing and they work great.  I wasn't able to test first letter navigation effectively for the Action menu because all the options start with the letter 'A' but moving through the options using all the methods seemed to work great.

**Comment by mcking65** [comment]
Earlier we agreed on the following text for the HCM focus indicator documentation:

>Because transparent borders are visible on some systems with high contrast enabled, only the focused element has a visible border. When focusable elements are not focused, they have a 0-width border and additional padding equal in width to the border that is used to indicate focus.

Now the text is:

>To make focus changes visible in high contrast settings, the menu button border width is changed from 1 pixel to 3 pixels . A corresponding adjustment in the padding is also made for the overall dimensions of the menu button to remain the same. 

This new text doesn't explain how focus changes are made visible. It just says the border changes but not when or why. I assume the why is still because we can't just make the border transparent when an item does not have focus. I assume the change from 1 to 3 is when the element is focused.

So, when items do not have focus, why do they have the 1 pixel border? Is this because the color differences that make the shape of a button or menu item visible disappear in HCM and so to compensate we give them a border?

And, in HCM, is the only difference between a focused item and a an item that doesn't have focus the thickness of the border?

**Comment by mcking65** [comment]
@jongund @carmacleod 

I want to make sure our HCM documentation is clear before merging. We can then use this language as the model for other examples.

Earlier we agreed on the following text for the HCM focus indicator documentation:

>Because transparent borders are visible on some systems with high contrast enabled, only the focused element has a visible border. When focusable elements are not focused, they have a 0-width border and additional padding equal in width to the border that is used to indicate focus.

At some point, the text was changed to:

>To make focus changes visible in high contrast settings, the menu button border width is changed from 1 pixel to 3 pixels . A corresponding adjustment in the padding is also made for the overall dimensions of the menu button to remain the same. 

This text doesn't explain how focus changes are made visible. It just says the border changes but not when or why. I assume the why is still because we can't just make the border transparent when an item does not have focus. I assume the change from 1 to 3 is when the element is focused.

So, when items do not have focus, why do they have the 1 pixel border? Is this because the color differences that make the shape of a button or menu item visible disappear in HCM and so to compensate we give them a border?

And, in HCM, is the only difference between a focused item and a an item that doesn't have focus the thickness of the border?

Guessing at answers, I propose the following text.

>Because transparent borders are visible on some systems with operating system high contrast settings enabled, transparency cannot be used to create a visual difference between the element that is focused an other elements. Instead of using transparency, the focused element has a thicker border and less padding. When an element receives focus, its border changes from 1 to 3 pixels and padding is reduced by 2 pixels. When an element loses focus, its border changes from 3 to 1 pixels and padding is increased by 2 pixels.

The above text does not explain why we have a 1 pixel border on elements that ar not focused. Taking a wild guess at what things look like to you sightees, should we add a bullet like:

>Because enabling operating system high contrast settings can remove color variations used to distinguish buttons and menu items from plain text, focusable elements are given a 1 pixel border.

**Comment by mcking65** [comment]
@spectranaut will resolve merge conflicts tomorrow (Oct 22).

Then, as soon as we have agreement on the HCM text mentioned in above comment, I will update the accessibility features section, mark editorial review complete,  and then merge.

**Comment by charmarkk** [comment]
> Guessing at answers, I propose the following text.
> 
> > Because transparent borders are visible on some systems with operating system high contrast settings enabled, transparency cannot be used to create a visual difference between the element that is focused an other elements. Instead of using transparency, the focused element has a thicker border and less padding. When an element receives focus, its border changes from 1 to 3 pixels and padding is reduced by 2 pixels. When an element loses focus, its border changes from 3 to 1 pixels and padding is increased by 2 pixels.
> 
> The above text does not explain why we have a 1 pixel border on elements that ar not focused. Taking a wild guess at what things look like to you sightees, should we add a bullet like:
> 
> > Because enabling operating system high contrast settings can remove color variations used to distinguish buttons and menu items from plain text, focusable elements are given a 1 pixel border.

I think this is a good combo/compromise, Matt!

**Comment by spectranaut** [comment]
All conflicts were related to new lint and prettier rules! Fixed now :)

**Comment by jongund** [comment]
@mcking65 
Javascript now uses class instead of prototype for defining widget object inline with new authoring practices.

**Comment by jongund** [comment]
@mcking65 
I have updated the high contrast documentation.

**Comment by mcking65** [comment]
As anticipated, we have some conflicts from recent merges. @jongund, if you could help resolve, that'd be great. Otherwise, Valerie can help when she has another day on APG.

This looks pretty much ready to go. I want to make some minor grammar edits to the accessibility features documentation. Then, I'll merge.

File: examples/menu-button/menu-button-actions-active-descendant.html
Comment (nschonni) [inline examples/menu-button/menu-button-actions-active-descendant.html:334]:
```suggestion
        </ul>
      </section>
```

**Comment by jongund** [comment]
@mcking65 
I fixed the conflicts, so should be ready for your edits.

**Review by mcking65 — APPROVED** [review]
Editorial review complete; ready to merge!

**Comment by mcking65** [comment]
@jongund  thank you, thank you, thank you for all the hard work put into this!!!!

**Pull request merged**
